### PR TITLE
(DOCSP-43057) Connect to MongoDB > Create a MongoClient

### PR DESCRIPTION
**Pull request by amalhotra-mdb**
# Pull Request Info

Create a child page of Connect to Mongo tree that instructs how to create a MongoClient. Since root page of Connect to Mongo is reliant on child page content + contains child page links, created stub for the root page in order to create the child page (will fill out stub in [this ticket](https://jira.mongodb.org/browse/DOCSP-43745))

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-43057>
Staging - <https://deploy-preview-43--docs-c.netlify.app/connect/mongoclient/>

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?
- [ ] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/connect/mongoclient.txt [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+For more information about creating a connection string, see
+:manual:`Connection Strings </reference/connection-string>` in the
+MongoDB Server documentation.
+
+Atlas Connection Example
+------------------------
+
+To connect to a MongoDB deployment on Atlas, you must first create a client. 
+
+You can pass a connection URI as a string to the ``mongoc_client_new()`` method
```
Comment (mcmorisi) [inline source/connect/mongoclient.txt:86]: Nit: C doesn't have methods, only functions
```suggestion
You can pass a connection URI as a string to the ``mongoc_client_new()`` function 
```

**Review by mcmorisi — CHANGES_REQUESTED** [review]
Good work! Just a handful of C-specific callouts.

File: source/connect/mongoclient.txt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+   :copyable:
+   :dedent:
+
+API Documentation
+-----------------
+
+For more information about creating a ``mongoc_client_t`` structure with the 
+{+driver-short+}, see the following API documentation: 
+
+- `mongoc_client_t <https://mongoc.org/libmongoc/current/mongoc_client_t.html>`__
```
Comment (mcmorisi) [inline source/connect/mongoclient.txt:120]: I think we can shorten this to just:
```suggestion
For more information about the ``mongoc_client_t`` structure, see the `API documentation <https://mongoc.org/libmongoc/current/mongoc_client_t.html>`__.

**Review by mcmorisi — APPROVED** [review]
LGTM with one last small comment!

File: source/connect.txt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+.. facet::
+   :name: genre
+   :values: reference
+ 
+.. meta::
+   :description: Learn how to use the C driver to connect to MongoDB.
```
Comment (eramongodb) [inline source/connect.txt:18]:
```suggestion
   :description: Learn how to use the C Driver to connect to MongoDB.
```
Should "Driver" be capitalized (proper noun)?

File: source/connect/mongoclient.txt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+     
+     - Optional. The authentication database to use if the
+       connection string includes ``username:password@``
+       authentication credentials but not the ``authSource`` option. If you don't include 
+       this component, the client authenticates the user against the ``admin`` database.
+
+   * - ``?<options>``
+
+     - Optional. A query string that specifies connection-specific
+       options as ``<name>=<value>`` pairs.
```
Comment (eramongodb) [inline source/connect/mongoclient.txt:75]:
```suggestion
       options as ``<name>=<value>`` pairs delimited by ``&``.
```
Disregard if this is considered too detailed (deferring to the Connection Strings docs).

File: source/includes/connect/mongoclient.c [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  // start-connect-to-atlas-w-uri
+  mongoc_client_t *client;
+  mongoc_database_t *database;
+  char *uri;
+
+  mongoc_init ();
+
+  // Create a new client and connect to the server
+  uri = "mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>";
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:14]: The standard connection string descrption only lists `mongodb://`, but the immediately following Atlas connection examples use `mongodb+srv://`. Recommend updating the connection string description to include SRV connection strings, which is what is used in example code.

File: source/includes/connect/mongoclient2.c [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  bson_error_t error;
+  char *uri;
+  bool retval;
+  bool ok = true;
+
+  // Intialize the MongoDB C Driver
+  mongoc_init();
+
+  // Create a new client and connect to the server
+  uri = "mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>";
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:19]:
```suggestion
  uri = "mongodb+srv://<db_username>:<db_password>@<hostname:port>/?<options>";
```
The `/` within `<>` is confusing; suggest using `:` (for hostname and port) or `|` to denote choice.

File: source/includes/connect/mongoclient.c [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#include <mongoc/mongoc.h>
+#include <bson/bson.h>
+
+int main (void) {
+
+  // start-connect-to-atlas-w-uri
+  mongoc_client_t *client;
+  mongoc_database_t *database;
+  char *uri;
+
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:10]:
```suggestion
```
The C Driver no longer supports C89/C90 compatibility, so mixing declarations and code is permitted and preferred.

```c
// Initialize the Mongo C Driver.
mongoc_init();

// Create a new client and connect to the server.
const char *uri = "...";
mongoc_client_t *client = mongoc_client_new (uri);
mongoc_database_t *database = mongoc_client_get_database (client, "admin");
```

File: source/includes/connect/mongoclient2.c [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  //start-connect-to-atlas
+  mongoc_client_t *client;
+  mongoc_server_api_t *api = NULL;
+  bson_t *command, reply;
+  bson_error_t error;
+  char *uri;
+  bool retval;
+  bool ok = true;
+
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:14]:
```suggestion
```
Avoid C89/C90 style declarations. See other comment regarding "C89/C90 compatibility".

File: source/includes/connect/mongoclient2.c [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  // Create a new client and connect to the server
+  uri = "mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>";
+  client = mongoc_client_new (uri);
+
+  // Set the version of the Stable API on the client
+  api = mongoc_server_api_new(MONGOC_SERVER_API_V1);
+  ok = mongoc_client_set_server_api(client, api, &error);
+
+  if (!ok) {
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:26]:
```suggestion
  if (!mongoc_client_set_server_api(client, api, &error)) {
```
Avoid redundant use-only-once local variables.

File: source/includes/connect/mongoclient2.c [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    fprintf(stderr, "error: %s\n", error.message);
+    return EXIT_FAILURE;
+  }
+    
+  // Send a ping to confirm a successful connection
+  command = BCON_NEW ("ping", BCON_INT32 (1));
+  retval = mongoc_client_command_simple (client, "admin", command, NULL, &reply, &error);
+
+  if (!retval) {
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:35]:
```suggestion
  if (!mongoc_client_command_simple (client, "admin", command, NULL, &reply, &error)) {
```
Avoid redundant use-only-once local variables.

File: source/includes/connect/mongoclient2.c [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  // Intialize the MongoDB C Driver
+  mongoc_init();
+
+  // Create a new client and connect to the server
+  uri = "mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>";
+  client = mongoc_client_new (uri);
+
+  // Set the version of the Stable API on the client
+  api = mongoc_server_api_new(MONGOC_SERVER_API_V1);
+  ok = mongoc_client_set_server_api(client, api, &error);
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:24]:
```suggestion
  {
     mongoc_server_api_t *api = mongoc_server_api_new (MONGOC_SERVER_API_V1);
     if (!mongoc_client_set_server_api (client, api, &error)) { ... }
     mongoc_server_api_destroy (api);
  }
```
Reduce scope of variables when able.

File: source/includes/connect/mongoclient2.c [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+  if (!ok) {
+    fprintf(stderr, "error: %s\n", error.message);
+    return EXIT_FAILURE;
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:28]: `return` risks skipping past cleanup routines, leading to memory leaks. Consider using a `goto failure;` or `goto cleanup;` pattern instead.

```c
if (...) {
  fprintf (stderr, ...);
  goto failure;
}

// ...

printf ("Pinged your deployment. You successfully connected to MongoDB!\n");

cleanup:
bson_destroy (&reply);
bson_destroy (commaned);
// ...
```

Reducing the scope of variables can also help reduce memory leak opportunities (cleanup before branch/jump).

File: source/includes/connect/mongoclient2.c (hunk 8 above)
Comment (amalhotra-mdb) [reply source/includes/connect/mongoclient2.c:14]: Thank you! Is it okay if `error` and `reply` are declared at the start of this code block, or can you think of a way to incorporate the declarations in the code?

Comment (eramongodb) [reply source/includes/connect/mongoclient2.c]: > Is it okay if [...] declared at the start of this code block?

That is fine. It is a common pattern to reuse an error object.

File: source/includes/connect/mongoclient2.c [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  // Set the version of the Stable API on the client
+  mongoc_server_api_t api = mongoc_server_api_new(MONGOC_SERVER_API_V1);
+
+  if (!mongoc_client_set_server_api (client, api, &error)) {
+    fprintf(stderr, "error: %s\n", error.message);
+    goto cleanup;
+  }
+    
+  // Send a ping to confirm a successful connection
+  bson_t *command = BCON_NEW ("ping", BCON_INT32 (1));
```
Comment (eramongodb) [inline source/includes/connect/mongoclient2.c:26]: Forgot to mention that goto mustn't jump over variable declarations, otherwise the variable is uninitialized during cleanup routines (undefined behavior).

Suggest modifying this example code to the following:

```c
#include <mongoc/mongoc.h>
#include <bson/bson.h>

int main(void) {
  //start-connect-to-atlas
  // Initialize the C Driver.
  mongoc_init();

  // Create a new client and connect to the server
  mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname:port>/?<options>");

  // Set the version of the Stable API on the client
  mongoc_server_api_t *api = mongoc_server_api_new(MONGOC_SERVER_API_V1);

  bson_t *ping = BCON_NEW ("ping", BCON_INT32 (1));
  bson_t reply = BSON_INITIALIZER;
  bson_error_t error;

  if (!mongoc_client_set_server_api (client, api, &error)) {
    fprintf(stderr, "error: %s\n", error.message);
    goto cleanup;
  }

  // Send a ping to confirm a successful connection
  if (!mongoc_client_command_simple (client, "admin", ping, NULL, &reply, &error)) {
    fprintf(stderr, "error: %s\n", error.message);
    goto cleanup;
  }

  printf ("Pinged your deployment. You successfully connected to MongoDB!\n");

cleanup:
  bson_destroy (&reply);
  bson_destroy (ping);

  mongoc_server_api_destroy(api);
  mongoc_client_destroy (client);

  // Cleanup the C Driver.
  mongoc_cleanup ();
  //end-connect-to-atlas
}
```

File: source/includes/connect/mongoclient.c [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+#include <bson/bson.h>
+
+int main (void) {
+
+  // start-connect-to-atlas-w-uri
+  mongoc_init ();
+
+  // Create a new client and connect to the server
+  char *uri = "mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>";
+  mongoc_client_t *client = mongoc_client_new (uri);
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:11]:
```suggestion
  mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname/port>/?<options>");
```
String literals are const. Recommend avoiding this detail by passing the URI string directly.

File: source/includes/connect/mongoclient.c [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+  mongoc_database_t *database = mongoc_client_get_database (client, "admin");
+
+  // Cleanup
+  mongoc_database_destroy (database);
+  mongoc_client_destroy (client);
+  mongoc_cleanup ();
+  // end-connect-to-atlas-w-uri
+}
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:19]:
```suggestion
}

```
C requires an EOF newline.

File: source/connect/mongoclient.txt [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+   :widths: 20 80
+   :header-rows: 1
+
+   * - Component
+     - Description
+
+   * - ``mongodb+srv://``
+
+     - Required. A prefix that identifies this as a string in the
+       standard connection format.
```
Comment (eramongodb) [inline source/connect/mongoclient.txt:53]:
```suggestion
       SRV connection format.
```

File: source/includes/connect/mongoclient.c [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#include <mongoc/mongoc.h>
+#include <bson/bson.h>
+
+int main (void) {
+
+  // start-connect-to-atlas-w-uri
+  mongoc_init ();
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:7]:
```suggestion
  // Initialize the C Driver.
  mongoc_init ();
```
For consistency.

File: source/includes/connect/mongoclient.c [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+  // Create a new client and connect to the server
+  mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname:port>/?<options>");
+  mongoc_database_t *database = mongoc_client_get_database (client, "admin");
+
+  // Cleanup
+  mongoc_database_destroy (database);
+  mongoc_client_destroy (client);
+  mongoc_cleanup ();
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:16]:
```suggestion

  // Cleanup the C Driver.
  mongoc_cleanup ();
```

For consistency.

File: source/includes/connect/mongoclient.c [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+int main (void) {
+
+  // start-connect-to-atlas-w-uri
+  mongoc_init ();
+
+  // Create a new client and connect to the server
+  mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname:port>/?<options>");
+  mongoc_database_t *database = mongoc_client_get_database (client, "admin");
+
```
Comment (eramongodb) [inline source/includes/connect/mongoclient.c:12]:
```suggestion

   // Use `database` here.

```

May want to insert a simple usage of the `database` variable here (e.g. a ping command) or mark the usage region with a comment (as suggested).

**Review by eramongodb — APPROVED** [review]
Minor feedback; otherwise, LGTM.
